src/main/java/com/mixpanel/android/mpmetrics/MixpanelPushNotification.java

```diff
@@ -320,18 +332,162 @@ protected void maybeAddActionButtons() {
-                mBuilder.addAction(this.createAction(btn.getIcon(), btn.getLabel(), btn.getUri()));
+                mBuilder.addAction(this.createAction(btn.getLabel(), btn.getOnTap(), btn.getId(), i + 1));
             }
         }
     }
 
-    private PendingIntent createActionIntent(String uri) {
-        return PendingIntent.getActivity(
-                mContext,
-                0,
-                new Intent(Intent.ACTION_VIEW, Uri.parse(uri)),
-                PendingIntent.FLAG_UPDATE_CURRENT
-        );
+    protected List<MixpanelNotificationData.MixpanelNotificationButtonData> buildButtons(String buttonsJsonStr) {
+        List<MixpanelNotificationData.MixpanelNotificationButtonData> buttons = new ArrayList<>();
+        if (null != buttonsJsonStr) {
+            try {
+                JSONArray buttonsArr = new JSONArray(buttonsJsonStr);
+                for (int i = 0; i < buttonsArr.length(); i++) {
+                    JSONObject buttonObj = buttonsArr.getJSONObject(i);
+
+                    // handle button label
+                    final String btnLabel = buttonObj.getString("lbl");
+
+                    // handle button action
+                    final MixpanelNotificationData.PushTapAction pushAction = buildOnTap(buttonObj.getString("ontap"));
+
+                    //handle button id
+                    final String btnId = buttonObj.getString("id");
+
+                    if (pushAction == null || btnLabel == null || btnId == null) {
+                        MPLog.d(LOGTAG, "Null button data received. No buttons will be rendered.");
+                    } else {
+                        buttons.add(new MixpanelNotificationData.MixpanelNotificationButtonData(btnLabel, pushAction, btnId));
+                    }
+
+                }
+            } catch (JSONException e) {
+                MPLog.e(LOGTAG, "Exception parsing buttons payload", e);
+            }
+        }
+
+        return buttons;
+    }
+
+    protected MixpanelNotificationData.PushTapAction buildOnTap(String onTapStr) {
+        MixpanelNotificationData.PushTapAction onTap = null;
+        if (null != onTapStr) {
+            try {
+                final JSONObject onTapJSON = new JSONObject(onTapStr);
+                final String typeFromJSON = onTapJSON.getString("type");
+
+                if (!typeFromJSON.equals(MixpanelNotificationData.PushTapTarget.HOMESCREEN.getTarget())) {
+                    final String uriFromJSON = onTapJSON.getString("uri");
+                    onTap = new MixpanelNotificationData.PushTapAction(MixpanelNotificationData.PushTapTarget.fromString(typeFromJSON), uriFromJSON);
+                } else {
+                    onTap = new MixpanelNotificationData.PushTapAction(MixpanelNotificationData.PushTapTarget.fromString(typeFromJSON));
+                }
+
+                if (onTap.getActionType().getTarget().equals(MixpanelNotificationData.PushTapTarget.ERROR.getTarget())) {
+                    hasOnTapError = true;
```

Comment:
If an unknown target is passed, I think we should default it to open the HOMESCREEN (ie let the OS handle the tap) and still display it. We should log an error / warning in console, too. wdyt?

Reply:
I think this behavior makes the most sense